Chainguard Repository for Java (#3316)
chainguard-dev/edu: content/chainguard/libraries/overview.md

- No longer provide security fixes for packages built against the EOL version
- Continue to serve previously built packages

## Upstream fallback and controls

Chainguard Libraries support an optional protected upstream fallback, managed through the [Chainguard
Repository](/chainguard/chainguard-repository/overview/).

By default, the Chainguard library endpoints serve
only Chainguard-built packages. When the upstream fallback is enabled, upstream packages are
subject to additional security controls before being served.

> Note: When using `chainctl` to configure upstream fallback or cooldown duration, it can take up to 30 minutes for the repository changes to take effect.

### Enable the upstream repository

To enable or change upstream fallback configuration, use the [`chainctl
libraries entitlements`

> command](/chainguard/chainctl/chainctl-docs/chainctl_libraries_entitlements_create/). You cannot enable fallback via the Chainguard Console.

Reviewer: this should be a note that you cannot enable fallback or view upstream vs cg-built libraries int he Console for Python and Java currently.

it is available for Javascript though

- No longer provide security fixes for packages built against the EOL version
- Continue to serve previously built packages

## Upstream fallback and controls

> Chainguard Libraries support an optional built-in fallback managed through the [Chainguard

Reviewer: ```suggestion
Chainguard Libraries support an optional protected upstream fallback, managed through the [Chainguard
```

Repository](/chainguard/chainguard-repository/overview/): 

- Chainguard Libraries for JavaScript supports fallback to
the upstream npm Registry
- Chainguard Libraries for Java supports fallback to Maven Central

By default, the Chainguard library endpoints serve
only Chainguard-built packages. When the upstream fallback is enabled, upstream packages are
subject to additional security controls before being served.

> Note: When using `chainctl` to configure upstream fallback or cooldown duration, it can take up to 30 minutes for the repository changes to take effect.

### Enable the upstream repository

To enable or change upstream fallback configuration, use the [`chainctl
libraries entitlements`
command](/chainguard/chainctl/chainctl-docs/chainctl_libraries_entitlements_create/).

The following command creates or updates an entitlement to Chainguard Libraries

> for JavaScript, adds the npm upstream fallback policy, and configures a 7-day cooldown:

Reviewer: with the new libraries policies we will want to remove the --cooldown-days parameter and just note that setting the upstream policy will automatically set a default 7 day cooldown period.

> for JavaScript, adds the npm upstream fallback policy, and by default includes a 7-day cooldown:

Reviewer: ```suggestion
for JavaScript, and adds the npm upstream fallback policy. Enabling the upstream fallback will by default include a 7-day cooldown, which can be later configured.
```

```bash
chainctl libraries entitlements create --ecosystems=JAVASCRIPT --policy=CHAINGUARD_AND_UPSTREAM 
```

### Fallback options

The following options are available:
* **No upstream fallback (default)**: Only Chainguard-built packages are served.

> * **Upstream fallback enabled with cooldown**: Upstream packages are available after passing a cooldown period and malware scan. The same cooldown period is also enforced for Chainguard-built packages when the upstream repository is enabled, so dependency trees resolve consistently across both sources.

Reviewer: ```suggestion
* **Upstream fallback enabled**: Upstream packages are available after passing a configurable cooldown period and malware scan. The same cooldown period is enforced across Chainguard-built packages and upstream packages, so that dependency trees resolve consistently across both sources.
```

> #### Configuring the cooldown period

Reviewer: should this section go under "Cooldown period" below?

When upstream fallback is enabled, users with the Owner role can configure the cooldown with `chainctl`:

```bash
chainctl libraries entitlements create --ecosystems=JAVASCRIPT --policy=CHAINGUARD_AND_UPSTREAM --cooldown-days=3
```

> The default cooldown period is 7 days. Note that shorter cooldown periods increase the risk of pulling malicious or compromised upstream packages before the broader ecosystem can detect and report them.

Reviewer: ```suggestion
The default cooldown period is 7 days. The cooldown period provides an additional layer of defense on top of malware and greyware scanning, giving the broader security community time to surface threats that may not be immediately detectable. 
```

> > **Upstream fallback best practices**

Reviewer: I think we may be able to remove this note now that we are doing our own malware scanning before serving any upstream packages

> Upstream packages are proxied directly from npm or Maven Central, and are not rebuilt or authored by Chainguard as part of our Libraries product. The cooldown period and malware scanning provide a supplemental baseline of protection to your own security practices, but you are solely responsible for independently evaluating and validating all upstream artifacts before use in your environment.

### Malware and greyware detection

Chainguard's [source code and maintainer behavior
scanning](https://www.chainguard.dev/unchained/the-expanding-threat-landscape-chainguard-now-scans-source-code-for-traditional-malware-and-greyware/)
identifies and blocks malicious and greyware packages in Chainguard Libraries
via the Chainguard Repository. This includes packages that are publicly reported
as malicious (including packages associated with OSV malware IDs) and packages

> that Chainguard determines are unsafe, even when no public malware advisory

Reviewer: ```suggestion
that Chainguard determines are unsafe through its own malware source code scanning, even when no public malware advisory
```

exists yet. If a package is flagged as malicious, Chainguard does not build that

> package from source or serve it via upstream fallback through the Chainguard

Reviewer: the "does not build that package from source" has not been implemented yet

The pull token is long-lived (configurable with `--ttl`, with a maximum lifetime of 1 year) and doesn't require a browser login, making it suitable for automation. 

{{< /details >}}

Next, use the malware API endpoint to return a list of libraries currently blocked due to malware. 

For example, to list all blocked JavaScript libraries from malware detection since May 1, 2026:

```bash
curl -H "Authorization: Bearer $(chainctl auth token --audience=libraries.cgr.dev)" "https://libraries.cgr.dev/javascript/-/api/malware?since=2026-05-01T00:00:00Z" | jq .
```

### Cooldown period

When fallback is enabled, upstream packages are subject to a cooldown period from their publication date before the Chainguard Repository will serve them. The cooldown is an additional layer of security that provides a window for the security community to identify and report malicious packages before your builds can pull them. 

The cooldown applies globally across Chainguard-built packages and upstream npm or Maven Central packages served through the fallback. This prevents installs from failing when a Chainguard-built package depends on an upstream dependency that is still under the cooldown window.

> If a package version is requested and falls within the cooldown period, the package manager will output a 404 error. The package becomes available once it has passed the cooldown period and cleared malware scanning.

Reviewer 2: ```suggestion
If a requested package version falls within the cooldown period, the package manager will output a 404 error. The package becomes available once it has passed the cooldown period and cleared malware scanning.
```

  environment to see if there are attempts to call out to an external server,
  read system files, or execute hidden payloads.

Use Chainguard's [malware API endpoint](/chainguard/api/spec-api-v1/#tag/malware) to query malware scanning details.

### Cooldown period

When fallback is enabled, upstream packages are subject to a cooldown period from their publication date before the Chainguard Repository will serve them. The cooldown is an additional layer of security that provides a window for the security community to identify and report malicious packages before your builds can pull them. 

The cooldown applies globally across Chainguard-built packages and upstream npm or Maven Central packages served through the fallback. This prevents installs from failing when a Chainguard-built package depends on an upstream dependency that is still under the cooldown window.

If a requested package version falls within the cooldown period, the package manager will output a 404 error. The package becomes available once it has passed the cooldown period and cleared malware scanning.

> > **Upstream fallback best practices**

Reviewer: same comment on removing this note - now that we are doing malware scanning acoss all usptream libraries, customers are able to rely on Chainguard for protection across all of them (similar to our chainguard-built libraries)

### How package resolution works

When you request a package from the Chainguard Repository, the following logic applies:

* **Chainguard-built package available**: The package is served directly from Chainguard's rebuilt artifact store, complete with SBOM, provenance, and signatures, subject to the configured cooldown.
* **Package not yet built by Chainguard**: If upstream fallback is enabled, the repository checks whether the package has passed the cooldown period and malware scan.
    * **Within the cooldown period**: The request returns an error. This prevents newly published packages — which carry higher malware risk — from being served immediately.
    * **After the cooldown period**: If upstream fallback is enabled and the version is outside the cooldown window and passes malware scanning, the repository pulls the version through from the upstream registry, serves it to the client, and caches it in the upstream mirror for future requests.

> * **Malware detected**: Any package version with a known malware identifier (MAL ID) is blocked and never served, whether it originates from Chainguard builds or the npm upstream. Malware scanning runs on all packages, including those proxied from npm.

Reviewer: ```suggestion
* **Malware or greyware detected**: Any package version that is detected for malware or greyware is blocked, whether it originates from Chainguard's builds or the upstream fallback.
```

Reviewer: this still refers to just MAL ID detection - i think we can make this brief and refer to the mawlare/greyware scanning section above. Also worth noting that overrides will be available for packages blocked under cooldown & detected as malware/greyware

The cooldown applies globally across Chainguard-built packages and upstream npm or Maven Central packages served through the fallback. This prevents installs from failing when a Chainguard-built package depends on an upstream dependency that is still under the cooldown window.

If a requested package version falls within the cooldown period, the package manager will output a 404 error. The package becomes available once it has passed the cooldown period and cleared malware scanning.

> **Upstream fallback best practices**
> Upstream packages are proxied directly from npm and are not rebuilt or authored by Chainguard as part of our Libraries product. The cooldown period and malware scanning provide a supplemental baseline of protection to your own security practices, but you are solely responsible for independently evaluating and validating all upstream artifacts before use in your environment.

### How package resolution works

> When you request a JavaScript or Java package from the Chainguard Repository, the following logic applies:

Reviewer: ```suggestion
When you request a package from the Chainguard Repository, the following logic applies:
```

Reviewer: might be easier rather than listing all the ecosystems, once we add Python + additional ecosystems?
